addons/binding/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/discovery/SleepIQBedDiscoveryParticipant.java

```diff
@@ -0,0 +1,90 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.sleepiq.discovery;
+
+import java.util.Map;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.sleepiq.SleepIQBindingConstants;
+import org.openhab.binding.sleepiq.config.SleepIQBedConfiguration;
+import org.openhab.binding.sleepiq.handler.SleepIQCloudHandler;
+import org.openhab.binding.sleepiq.handler.SleepIQDualBedHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.syphr.sleepiq.api.model.Bed;
+
+/**
+ * The {@link SleepIQBedDiscoveryParticipant} is responsible for processing the
+ * results of devices found through the SleepIQ cloud service.
+ *
+ * @author Gregory Moyer - Initial contribution
+ */
+public class SleepIQBedDiscoveryParticipant extends AbstractDiscoveryService {
+
+    private static final int TIMEOUT = 60;
+
+    private final Logger logger = LoggerFactory.getLogger(SleepIQBedDiscoveryParticipant.class);
+
+    private final SleepIQCloudHandler cloudHandler;
+
+    /**
+     * Discovers beds in the SleepIQ cloud service account.
+     *
+     * @param cloudHandler the cloud service handler (bridge)
+     */
+    public SleepIQBedDiscoveryParticipant(final SleepIQCloudHandler cloudHandler) {
+
+        super(SleepIQDualBedHandler.SUPPORTED_THING_TYPE_UIDS, TIMEOUT, true);
+        this.cloudHandler = cloudHandler;
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery for new beds");
+        startScan();
+    }
+
+    @Override
+    protected void startScan() {
+
+        logger.debug("Starting scan for new beds");
+
+        for (Bed bed : cloudHandler.getBeds()) {
+
+            // only dual chamber beds are supported currently
+            if (!bed.isDualSleep()) {
+                logger.info("Found a bed that is not dual chamber - currently unsupported");
+                continue;
+            }
+
+            ThingUID bridgeUID = cloudHandler.getThing().getUID();
+            ThingUID thingUID = new ThingUID(SleepIQBindingConstants.THING_TYPE_DUAL_BED, bridgeUID,
+                    bed.getMacAddress());
+
+            // thing already exists
+            if (cloudHandler.getThingByUID(thingUID) != null) {
+                continue;
+            }
+
+            logger.debug("New bed found with MAC address {}", bed.getMacAddress());
+
+            @SuppressWarnings({ "unchecked", "rawtypes" })
+            Map<String, Object> properties = (Map) cloudHandler.updateProperties(bed, null);
```

Comment:
The reason for the cast is subtle. I have two places where I need to convert the Bed object to a set of properties. One is here within the discovery participant and DiscoveryResultBuilder#withProperties requires a Map<String, **Object**>. The other is when initializing the SleepIQDualBedHandler where BaseThingHandler#editProperties returns a Map<String, **String**> and BaseThingHandler#updateProperties requires a Map<String, **String**>. Since String is more specific, the method SleepIQCloudHandler#updateProperties was made to take and return Map<String, **String**>. Without the cast, javac throws an error here even though it is safe.

As to passing a new HashMap into the method here or creating it inside the updateProperties method, I'm fine with switching it. I'll push updates shortly with this changed.